torch/csrc/jit/passes/peephole.cpp

```diff
@@ -15,188 +18,393 @@ static bool mustBeEqual(const c10::optional<T>& a, const c10::optional<T>& b) {
+          changed_ = true;
+        }
+      } else if (node->matches("prim::device(Tensor a) -> Device")) {
+        auto ptt = node->input()->type()->expect<TensorType>();
+        if (ptt->device()) {
+          WithInsertPoint guard(node);
+          auto output = node->owningGraph()->insertConstant(*ptt->device());
+          GRAPH_UPDATE(
+              "Replacing ",
+              *node,
+              " with a device constant ",
+              output->debugName());
+          node->output()->replaceAllUsesWith(output);
+          changed_ = true;
+        }
+      } else if (node->matches("aten::dim(Tensor self) -> int")) {
+        auto ptt = node->input()->type()->expect<TensorType>();
+        if (auto dim = ptt->sizes().size()) {
+          WithInsertPoint guard(node);
+          auto output =
+              node->owningGraph()->insertConstant(static_cast<int64_t>(*dim));
+          GRAPH_UPDATE(
+              "Replacing ",
+              *node,
+              " with a \"dim\" constant ",
+              output->debugName());
+          node->output()->replaceAllUsesWith(output);
+          changed_ = true;
+        }
+      } else if (node->matches("prim::is_cuda(Tensor a) -> bool")) {
+        auto ptt = node->input()->type()->expect<TensorType>();
+        if (ptt->device()) {
+          WithInsertPoint guard(node);
+          auto output =
+              node->owningGraph()->insertConstant((*ptt->device()).is_cuda());
+          GRAPH_UPDATE(
+              "Replacing ",
+              *node,
+              " with a is_cuda constant ",
+              output->debugName());
+          node->output()->replaceAllUsesWith(output);
+          changed_ = true;
+        }
       }
-      // TODO: this doesn't work with Scalar-Tensor ops! We should canonicalize
-      // those
-    } else if (
-        node->matches(
-            "aten::mul(Tensor self, Scalar other) -> Tensor",
-            /*const_inputs=*/attr::other) ||
-        node->matches(
-            "aten::div(Tensor self, Scalar other) -> Tensor",
-            /*const_inputs=*/attr::other)) {
-      // x * 1 == x / 1 == x
-      if (node->get<at::Scalar>(attr::other)->toDouble() == 1) {
-        GRAPH_UPDATE(
-            *node,
-            " (x * 1 == x / 1 == x) is replaced with ",
-            node->input(0)->debugName());
-        node->output()->replaceAllUsesWith(node->input(0));
-      }
-    } else if (
-        node->matches(
+
+      // these transformations rely on AA for correctness
+      // see `runAliasingSensitivePeepholeTransformations` for more details
+      runAliasingSensitivePeepholeTransformations(node);
+    }
+  }
+
+  // if either the inputs or outputs of an op alias graph's inputs or
+  // outputs, the transformations below are invalid
+  // An example:
+  //
+  // def test_write(x):
+  //     s = 0
+  //     s += x
+  //     s += x
+  //     return s
+  //
+  void runAliasingSensitivePeepholeTransformations(Node* node) {
+    if (changed_) {
```

Comment:
Can you move these 10 lines out to a function and only call it within the body of `node->matches` write before the `node->output()->replaceAllUsesWith` logic? 

We don't need to run this query for every node in the graph